dev-guide/api-conventions.md

```diff
@@ -129,6 +129,139 @@ By providing quality documentation within the API itself, a number of generated
+We do not want our code to guess what the user meant, they should tell us which of the choices they made using the discriminant.
+
+#### Writing a union in Go
+
+Union types in Go have some additional helper tags which signify how the structure should be handled to consumers.
+Below is an example based on the idea of platform types.
+
+```go
+// MyPlatformConfig is a discriminated union of platform specific configuration.
+// It has a +union tag which informs consumers that this is expected to be a union type.
+// +union
+type MyPlatformConfig struct {
+  // PlatformType is the unions discriminator.
+  // Users are expected to set this value to the name of the platform.
+  // The value of this field should match, or be able to be mapped to exactly one field
+  // in the union structure.
+  // It has a +unionDiscriminator tag to inform consumers that this is the discriminator field.
+  // The field should be an enum type, so you may also need an enum tag.
+  // The field is required, though a future KEP will default this field if not set, at which point
+  // it can be made optional. See future section below.
+  // +unionDiscriminator
+  // +kubebuilder:validation:Enum:="AWS";"Azure";"GCP"
+  // +kubebuilder:validation:Required
+  PlatformType string `json:"platformType,omitempty"`
+
+  // AWS is the AWS configuration.
+  // All structures within the union must be optional and pointers.
+  // +optional.
+  AWS *MyAWSConfig `json:"aws,omitempty"`
+
+  // Azure is the Azure configuration.
+  // All structures within the union must be optional and pointers.
+  // +optional.
+  Azure *MyAzureConfig `json:"azure,omitempty"`
+
+  // GCP is the GCP configuration.
+  // All structures within the union must be optional and pointers.
+  // +optional.
+  GCP *MyGCPConfig `json:"gcp,omitempty"`
+}
+```
+
+The discriminator here allows the consumer to determine which of the configuration structures they should be consuming, AWS, Azure or GCP.
+
+Important to note:
+* All structs within the union **MUST** be pointers
+* All structs within the union **MUST** be optional
+* The discriminant should be required, but will be defaulted by the API server in a future Kubernetes version, at which point it can be optional
+* The discriminant **MUST** be a string (or string alias) type
+* If a discriminant value does not match the json tag of the struct to which it refers, a mapping can be made using an additional tag.
+
+#### Using union types
+
+Below are some examples of how a user may configure the above example.
+
+##### Case 1
+
+```yaml
+myPlatformConfig:
+  platformType: AWS
+  aws:
+    ...
+```
+
+This is valid. Only one struct is configured and the discriminant is correct.
+
+##### Case 2
+
+```yaml
+myPlatformConfig:
+  platformType: AWS
+  aws:
+    ...
+  azure:
+    ...
+```
+
+This is invalid. The Azure configuration should not be configured when the `platformType` is AWS.
+
+In the future, the API server will silently drop the Azure configuration.
```

Comment:
Let's avoid future guarantees, I'm more comfortable with a failure on the Azure field.  We'll see how it lands.